### ISSUE TITLE
Experiment module: Link geometry and variable upload files to revision history

### ISSUE THREAD

**Issue by paul121** [Rothamsted-Ecoinformatics/farm_rothamsted#645]
Right now it is not easy to determine the "first upload" of column, plot variable and geojson data. These fields are dumped the `plan.file` field and have data parsed out into JSON fields on the Plan and Plot entities, but they are separate.

Being able to determine the "first upload" would benefit us in a few ways: https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues/632#issuecomment-2043154398

I would like to create dedicated file fields on the plan for column + column_level + plot_attribute CSVs + plot geojson. This would allow us to look in a single place to see if this data has been uploaded in a "first upload". It would also make it easier to determine what the "current set" of data files are for the plan, rather than parsing through a longer list of files that have been uploaded and attached to the plan.

**Comment by aislinnpearson** [comment]
@mstenta, @paul121 and I had an interesting conversation about this issue. We realised that it overlaps with a previous conversation we had were FarmOS is increasingly becoming the source of truth at Rothamsted, and so if more than one version of a file is updated, it can be confusing as to which is the 'correct' or current version. At the time we suggested simply not keeping a copy of those files, but instead creating an 'export' button that would allow you to download the three .csv files and an annotated geoJSON (which also links to this issue here about Frictionless data exports #522).

_However_ in today's call we agreed this is not ideal: better to keep a copy of the file just in case. @paul121 came up with what is the ideal solution then that if you replace these named files (as per his comment above) they would exist only with that revision. In this way you keep a copy of all the files, clearly associated with the current revision. I think that is a nice solution to implement asap.

In terms of the longer term/ bigger picture we talked about how files are handled more generally and I have created a separate issue for that here:

**Comment by aislinnpearson** [comment]
Agreed in today's call that the main limitation is the database update hook/ migration for the existing files, and then see how many would be left to do manually.

We can then test this on the database associated with the .dev site to do an initial check, possibly followed by a check by Mike on the  production site.

**Comment by paul121** [comment]
Okay, here is a query we can run to try and give us a better idea what we are working with: https://gist.github.com/paul121/bef27d04a1eab16b90e3bcea2926074d @mstenta could you run this on production sites when you have a chance?

On my local dev snapshot I find 98 plans that have column descriptors in the database, meaning they should have at least some of these files we're talking about migrating.

Of these, 43 have "multiple files" uploaded: meaning there are more than 3 total files uploaded, and at least one of these files ends with `_0.csv`, `_1.csv`, `_2.csv`, etc, likely indicating that a newer version of columns, levels, attributes, or plot CSVs has been uploaded and will need a "manual" migration to determine which are the latest file versions.

There are 25 "good" plans, meaning each of these only has 3 files, all named `columns.csv`, `column-levels.csv`, or `plot-attributes.csv`. These plans can be migrated "automatically".

25+43=68. These 30 remaining plans, I'm not sure... they could possibly have "multiple files" or just be in a an intermediate state. It's very possible I have some random test data too, but likely not 30 plans.

**Comment by mstenta** [comment]
> @mstenta could you run this on production sites when you have a chance?

Done! I emailed you both a copy of the results.

**Comment by aislinnpearson** [comment]
Hey @paul121,

I think this sounds manageable - just need to talk through today what I need to do, and maybe test that on the dev site first?

**Comment by aislinnpearson** [comment]
Agreed today that there is no other solution than to go manually plan by plan and select the three or four files that are the current uploads, and link them to the new fields. @paul121 will create a form for this, and we can test it on the dev site first to make sure it's working correctly. This should be a relatively small job for the data admin as only 260 plans.

The revision message associated with the form should ideally be pre-filled but editable so additional comments can be added.

@paul121 how does this sound for a revision message: 

"This revision was created as part of the data migration for Github issue 645 (https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues/645). As part of this data migration, only the most recent csv and geoJSON files used to create the column descriptors and plot attributes are shown in the current plan. Where multiple files have been uploaded (for example, if the plot attributes or geometry have been updated) the old versions are archived but remain available the earlier revisions of the plan."

**Comment by aislinnpearson** [comment]
Hey @paul121 

Here is what I tested:

1. Selecting 'No action' results in files remaining under files **works as expected**
2. Selecting 'Archive' removes files from the current revision (but they remain in the previous revisions) **works as expected**
3. Selecting 'Columns' labels the file as 'Columns' in the new revision **works as expected**
4. Selecting 'Column levels' labels the file as 'Column levels' in the new revision **works as expected**
5. Selecting 'Plot attributes' labels the file as 'Plot attributes' in the new revision **works as expected**
6. Selecting 'Plot geometries' labels the file as 'Plot geometries' in the new revision **works as expected**
7. Uploading variables to an existing plan creates the new data fields **works as expected**
8. Uploading geometries to an existing plans creates the new data fields **works as expected by see note on revisions below**
9. Uploading variables to an existing plan creates the new data fields **works as expected**
10.  Uploading geometries to an existing plans creates the new data fields **works as expected by see note on revisions below**
11. Fields are created when setting up a new plan **works as expected by see note on revisions below**

There was some additional functionality that I noticed you added, which I really like:

- Where previously you could delete these files from the Plan, that is no longer possible. You can only re-upload them, which to me is correct _but_ I wonder what we should do about clearing the variables (for example, if they are all wrong). This reminds me of a related issue of if you get the number of plots wrong when creating them, you then have to delete the whole plan and start again. **Perhaps we combine these two problems and add this as a separate issue to address?**

I then started playing around with the revisions and noticed a few issues

1. **uploading new geometries _doesn't_ create a new revision:** when I re-upload the plot geometries, it **doesn't** create a new revision. I  tested this twice, and you can see the file updated, now labelled as an 'updated' geojson' which is the name of the file I uploaded) but the revisions aren't there (badly drawn errors indicating where they should exist for both my attempts):

<img width="947" height="209" alt="Image" src="https://github.com/user-attachments/assets/04063b1f-a716-4582-8b0e-1614ca8105d3" />

<img width="1432" height="433" alt="Image" src="https://github.com/user-attachments/assets/2f0f75ca-375d-4910-b96e-de7d256a598a" />

2. **reverting to previous versions creates problems:** This functionality works as I would expect it to in that it becomes an exact copy of the previous revision _except_ you can't actually access the previous files. If I click on a file in the revisions, it takes me to a page like the below. As the files don't exist, I get the same if I revert from a previous revision (newly uploaded files _do_ download). However, I don't know if that is only because this is a test site. If I remember correctly, the files weren't copied by @mstenta when we duplicated the database. From a research QA perspective, ideally we **would** still be able to access the files in the previous revisions, so I just wanted to double check (see here: https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues?q=is%3Aissue%20state%3Aclosed%20database). That said, the fact that reverting to a previous version undoes the migration step isn't ideal - I think this is yet another reason why we _wouldn't_ want people to be able to revert to previous versions and instead update the fields manually.

<img width="1263" height="553" alt="Image" src="https://github.com/user-attachments/assets/29656d60-d105-46ae-b914-f87077089261" />

3. **The timestamp associated with the revisions doesn't change**: It' just says the same time repeatedly, even though I was doing this over about half an hour.

<img width="854" height="700" alt="Image" src="https://github.com/user-attachments/assets/5ef46904-054c-48eb-9d53-c4ad7a1a4033" />

**Comment by paul121** [comment]
@aislinnpearson some responses to your questions:

1. Great find - I've fixed this and standardized how revisions are saved on all plan forms (plot, geometry and variables)
2. Yes - the files on the demo site are expected to be "missing". But we tested that new files are working. And we agreed that "reverting" plans for this use-case shouldn't be a normal thing.
3. I've fixed this while standardizing for # 1, should be working great.

**Comment by paul121** [comment]
I'm going to close - I think this is all working as expected. We can use a new issue for any followups

**Comment by aislinnpearson** [comment]
Note that for plans where the plot attributes were uploaded with the geoJSON, these were assigned to the plot geometry file field, and there is no plot attributes csv assigned to the plot attributes file field. 

See here for an example of this: https://rothamstedfarm.farmos.net/plan/15

<img width="1036" height="274" alt="Image" src="https://github.com/user-attachments/assets/0effa54a-c028-4b2b-a4f9-d55381786997" />
